### PR TITLE
Cheaper worker coordination

### PR DESCRIPTION
**Pull request by melix**
This pull request implemets cheaper worker coordination by using finer grained locks. This work is more important as we increase the parallelism "by default" in Gradle.

### CODE REVIEW

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/OutputEventRenderer.java [hunk 1]
```diff
@@ -290,30 +292,30 @@ public void configureMaxWorkerCount(int maxWorkerCount) {
...
+            this.logLevel.set(newLogLevel);
+        } else if (event instanceof MaxWorkerCountChangeEvent) {
+            MaxWorkerCountChangeEvent changeEvent = (MaxWorkerCountChangeEvent) event;
+            int newMaxWorkerCount = changeEvent.getNewMaxWorkerCount();
+            if (newMaxWorkerCount == this.maxWorkerCount.get()) {
+                return;
             }
+            this.maxWorkerCount.set(newMaxWorkerCount);
+        }
+        synchronized (lock) {
```
Comment (oehme) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/OutputEventRenderer.java]: ⭕️  Could this be in an else branch or do the formatters need to know about level/max worker change events?

Comment (melix) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/OutputEventRenderer.java]: I don't know. I'm not even sure that we need to lock at all, here.

File: subprojects/base-services/src/main/java/org/gradle/internal/operations/DefaultBuildOperationWorkerRegistry.java [hunk 2]
```diff
@@ -150,9 +156,6 @@ void releaseLease() {
             children--;
             if (children > 0) {
                 root.releaseLease();
-                lock.notify();
```
Comment (oehme) [inline subprojects/base-services/src/main/java/org/gradle/internal/operations/DefaultBuildOperationWorkerRegistry.java]: push -f? :)

Comment (melix) [inline subprojects/base-services/src/main/java/org/gradle/internal/operations/DefaultBuildOperationWorkerRegistry.java]: sorry?

Comment (oehme) [inline subprojects/base-services/src/main/java/org/gradle/internal/operations/DefaultBuildOperationWorkerRegistry.java]: Sorry I missed your revert, it looked like my commit was just gone :)

**Comment by melix** [comment]
The problem is that I'm blocked if I don't do this change. It kills performance for the parallel metadata case.

**Comment by melix** [comment]
I have cherry picked all changes that do not impact `DefaultBuildOperationWorkerRegistry` onto `master`.

**Pull request closed without merging**